### DISCUSSION TITLE
Where in .sol files should we declare errors, and why ?

### DISCUSSION THREAD

**Discussion by cebidhem** [Cyfrin/foundry-full-course-cu#741] [Q&A] [answered]
Hi there 👋 

I have a question on error declarations, and I'm not sure I fully understand why we may decide to declare them inside or outside of a `contract ContractName {}`.

In Lesson 9, Patrick is using the Solidity layout to build the contract, with the following order:

Layout of Contract:
- version
- imports
- errors
- interfaces, libraries, contracts
- Type declarations
- State variables
- Events
- Modifiers
- Functions

In the above order, errors are meant to be declared before the contract. 
This is what is done in [Lesson 7 FundMe.sol](https://github.com/Cyfrin/foundry-fund-me-f23/blob/main/src/FundMe.sol#L10)):
```solidity
// SPDX-License-Identifier: MIT
// 1. Pragma
pragma solidity ^0.8.19;
// 2. Imports

import {AggregatorV3Interface} from "@chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";
import {PriceConverter} from "./PriceConverter.sol";

// 3. Interfaces, Libraries, Contracts
error FundMe__NotOwner();

/**
 * @title A sample Funding Contract
 * @author Patrick Collins
 * @notice This contract is for creating a sample funding contract
 * @dev This implements price feeds as our library
 */
contract FundMe {}
```
However in [Lesson 9 Raffle.sol](https://github.com/Cyfrin/foundry-smart-contract-lottery-f23/blob/main/src/Raffle.sol#L37), errors are all declared within the contract:
```solidity
// SPDX-License-Identifier: MIT

pragma solidity ^0.8.19;

import {VRFCoordinatorV2Interface} from "@chainlink/contracts/src/v0.8/interfaces/VRFCoordinatorV2Interface.sol";
import {VRFConsumerBaseV2} from "@chainlink/contracts/src/v0.8/VRFConsumerBaseV2.sol";
import {AutomationCompatibleInterface} from "@chainlink/contracts/src/v0.8/interfaces/AutomationCompatibleInterface.sol";

/**@title A sample Raffle Contract
 * @author Patrick Collins
 * @notice This contract is for creating a sample raffle contract
 * @dev This implements the Chainlink VRF Version 2
 */
contract Raffle is VRFConsumerBaseV2, AutomationCompatibleInterface {
    /* Errors */
    error Raffle__UpkeepNotNeeded(
        uint256 currentBalance,
        uint256 numPlayers,
        uint256 raffleState
    );
    error Raffle__TransferFailed();
    error Raffle__SendMoreToEnterRaffle();
    error Raffle__RaffleNotOpen();
}
```

I guess I'm trying to understand what is the impact of declaring it inside or outside, and what's the convention here ?

I think it defines the scope of the errors, but basically is the convention to declare them inside a contract when there's only a contract in the file, and outside when it may be used by interfaces/libraries written in the file as well?

Thanks a lot for your help 🙏

**Comment by Naman1729** [comment]
Hey @cebidhem , if we declare custom Error inside the contract then we can fetch the error in the test to check which error is reverting by the contract like this :

```solidity
 vm.expectRevert(Raffle.Raffle__NotEnoughETHEntered.selector);
 ```
 on the other hand, if we declare custom Error outside the contract then we can not make sure which error is reverting by contract and we test like this :
 
 ```solidity
 vm.expectRevert();
 ```

**Reply by cebidhem** [reply]
Hi @Naman1729 ,

Thanks for your answer!

If I understand correctly, declaring those errors inside or outside the contracts really only impact the test capabilities, is that right ?

If so, I guess the practice is to always declare those errors in the contract right ?

**Reply by Naman1729** [reply]
@cebidhem, In my opinion custom error inside the contract is good practice👍

**Comment by itznishant** [comment, marked as answer]
> 

Custom errors are a best practise and recommended approach in solidity based smart contracts to help with debugging.
A good practise to declare any custom errors right at the start **inside** the contract declaration.
You should declare custom errors the contract below,  I have coded this ```Example.sol``` contract to help explain this better. 

![examplesolsnap](https://github.com/Cyfrin/foundry-full-course-f23/assets/17523443/1131c298-bbb1-4c36-b321-b47ed8861e6b)

Notice I have used multiple custom errors in ```addNumber``` function above to help the developer to pin point the location where the code failed with the specific custom error revert.

**Case 1:** If any inputs are greater than a target value, the code reverts right there with that specific custom error.
<img width="541" alt="custom__error_a" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/17523443/f3c61509-f0aa-4366-9929-a6621164cfe2">

The dev can then immediately understand that's this was an issue with one of the incorrectly supplied inputs.

**Case2:** When the inputs are within threshold, but the result is above an expected target. The ```addNumber``` function would revert for inputs { a=10, b=10 }

The custom error ```Example__ResultExceedsThreshold()``` clearly identifies this for efficient and fast debugging. 
<img width="541" alt="custom__error_result" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/17523443/28d9bc25-4459-425a-ab91-4072fd6a10fb">

Note: Above revert screenshots are taken from Remix IDE console output.

Hope this post help answer your query.

Do mark this as an answer as this should clearly help understand where and why do we need to add such custom errors in our smart contracts.

**Reply by itznishant** [reply]
@cebidhem see above reply.
